lib/seahorn/FiniteMapTransf.cc

```diff
@@ -0,0 +1,350 @@
+#include "seahorn/FiniteMapTransf.hh"
+
+#include "seahorn/Expr/ExprCore.hh"
+#include "seahorn/Expr/ExprOpBind.hh"
+#include "seahorn/Expr/ExprVisitor.hh"
+
+#include "seahorn/Support/SeaDebug.h"
+
+using namespace expr;
+using namespace expr::op;
+
+namespace seahorn {
+// ----------------------------------------------------------------------
+//  FiniteMapArgsVisitor
+// ----------------------------------------------------------------------
+
+// -- creates a constant with the name get(map,k)
+// also used in FinteMapBodyVisitor
+static Expr mkVarGet(Expr mapConst, Expr k, Expr vTy) {
+  assert(bind::isFiniteMapConst(mapConst));
+  return bind::mkConst(variant::variant(0, finite_map::get(mapConst, k)), vTy);
+}
+
+// \brief rewrites a map into separate scalar variables. New variables are added
+// to `new_vars`, new unifications are added to `extra_unifs`
+void mkVarsMap(Expr map, const ExprVector &keys, Expr vTy, ExprVector &new_vars,
+               ExprVector &extra_unifs, ExprSet &evars) {
+
+  Expr v, v_get;
+  ExprVector map_values(keys.size());
+  auto val_it = map_values.begin();
+
+  for (auto k : keys) {
+    v = mkVarGet(map, k, vTy);
+    evars.insert(v);
+    new_vars.push_back(k);
+    new_vars.push_back(v);
+    *val_it++ = v;
+  }
+  extra_unifs.push_back(
+      mk<EQ>(map, finite_map::constFiniteMap(keys, map_values)));
+}
+
+// \brief rewrites the map arguments of fapps into separate scalar variables
+static Expr mkFappArgsCore(Expr fapp, Expr newFdecl, ExprVector &extraUnifs,
+                           ExprSet &evars, ExprFactory &efac) {
+
+  assert(isOpX<FAPP>(fapp));
+
+  Expr fdecl = bind::name(fapp);
+  assert(bind::isFdecl(fdecl));
+  Expr fname = bind::fname(fdecl);
+
+  ExprVector newArgs;
+
+  auto arg_it = ++(fapp->args_begin()), t_it = ++(fdecl->args_begin());
+  if (arg_it == fapp->args_end()) // no args
+    return fapp;
+
+  int arg_count = 0;
+  for (; arg_it != fapp->args_end(); arg_it++, arg_count++, t_it++) {
+    Expr arg = *arg_it;
+    Expr argTy = *t_it;
+
+    if (isOpX<FINITE_MAP_TY>(argTy)) {
+      Expr map_var_name;
+      if (bind::isFiniteMapConst(arg)) {
+        map_var_name = arg;
+      } else {
+        map_var_name = bind::mkConst(variant::variant(arg_count, fname), argTy);
+        // if there is no name, we create a variant with the name of the
+        // function, make new variable (same as normalization)
+      }
+      Expr ksTy = finite_map::keys(argTy);
+      ExprVector keys(ksTy->args_begin(), ksTy->args_end());
+      Expr lmdKeys = finite_map::mkKeys(keys, efac);
+
+      mkVarsMap(map_var_name, keys, finite_map::valTy(argTy), newArgs,
+                extraUnifs, evars);
+      // new arguments are added to `newArgs` in the function above
+    } else {
+      newArgs.push_back(arg);
+    }
+  }
+  return bind::fapp(newFdecl, newArgs); // building the new fapp
+}
+
+VisitAction FiniteMapArgsVisitor::operator()(Expr exp) {
```

Comment:
some more comments on what the code is doing would be very helpful. I'm trying to reconstruct the spec based on the implementation and not completely sure I get it right.